src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/Serializer/WritingPrimitives.cs

```diff
@@ -0,0 +1,312 @@
+        {
+            buffer[writePosition++] = (byte)(0x80 | (value & 0x7F));
+            value >>= 7;
+        }
+
+        if (writePosition < buffer.Length)
+        {
+            buffer[writePosition++] = (byte)value;
+            return writePosition;
+        }
+
+        // Handle case of insufficient buffer space.
+        while (value >= ULong128)
+        {
+            writePosition = WriteSingleByte(ref buffer, writePosition, (byte)((value & 0x7F) | 0x80));
+            value >>= 7;
+        }
+
+        writePosition = WriteSingleByte(ref buffer, writePosition, (byte)value);
+
+        return writePosition;
+    }
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    internal static int WriteInt64WithTag(ref byte[] buffer, int writePosition, int fieldNumber, ulong value)
+    {
+        writePosition = WriteTag(ref buffer, writePosition, fieldNumber, WireType.VARINT);
+        writePosition = WriteVarInt64(ref buffer, writePosition, value);
+
+        return writePosition;
+    }
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    internal static int WriteDoubleWithTag(ref byte[] buffer, int writePosition, int fieldNumber, double value)
+    {
+        writePosition = WriteTag(ref buffer, writePosition, fieldNumber, WireType.I64);
+        writePosition = WriteFixed64LittleEndianFormat(ref buffer, writePosition, (ulong)BitConverter.DoubleToInt64Bits(value));
+
+        return writePosition;
+    }
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    internal static int WriteStringWithTag(ref byte[] buffer, int writePosition, int fieldNumber, string value)
+    {
+#if NETFRAMEWORK || NETSTANDARD2_0
+        int numberOfUtf8CharsInString;
+        unsafe
+        {
+            fixed (char* strPtr = value)
+            {
+                numberOfUtf8CharsInString = Encoding.UTF8.GetByteCount(strPtr, value.Length);
+            }
+        }
+#else
+        int numberOfUtf8CharsInString = Encoding.UTF8.GetByteCount(value);
+#endif
+
+        writePosition = WriteTag(ref buffer, writePosition, fieldNumber, WireType.LEN);
+        writePosition = WriteLength(ref buffer, writePosition, numberOfUtf8CharsInString);
+        EnsureBufferCapacity(ref buffer, writePosition + numberOfUtf8CharsInString);
+
+#if NETFRAMEWORK || NETSTANDARD2_0
+        _ = Utf8Encoding.GetBytes(value, 0, value.Length, buffer, writePosition);
+#else
+        _ = Encoding.UTF8.GetBytes(value, buffer.AsSpan().Slice(writePosition));
+#endif
+        writePosition += numberOfUtf8CharsInString;
+        return writePosition;
+    }
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    internal static int WriteSingleByte(ref byte[] buffer, int writePosition, byte value)
+    {
+        if (buffer.Length == writePosition)
+        {
+            if (!IncreaseBufferSize(ref buffer))
+            {
+                // TODO: throw an exception to indicate that the buffer is too large.
+            }
+        }
```

Comment:
I'm not going to drop a comment everywhere for this, but it might make sense to remove all of this expansion logic. Just assume the buffer has room. If not, catch indexoutofrange exception upstream, expand the buffer, and retry the entire serialization operation. This is what we do in prometheus. Why? Once we get to stable buffer size it will be a lot of faster without needing to do these checks each time we write some tiny piece of data into the blob. Also we can move from `ref byte[]` everywhere to just `byte[]`.